src/EditorFeatures/Core/Wrapping/AbstractParameterWrappingCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Editor.Wrapping
+{
+    internal abstract partial class AbstractWrappingCodeRefactoringProvider<
+        TListSyntax,
+        TListItemSyntax>
+        : CodeRefactoringProvider
+        where TListSyntax : SyntaxNode
+        where TListItemSyntax : SyntaxNode
+    {
+        private static ImmutableArray<string> s_mruTitles = ImmutableArray<string>.Empty;
+
+        protected abstract string ListName { get; }
+        protected abstract string ItemNamePlural { get; }
+        protected abstract string ItemNameSingular { get; }
+
+        protected abstract TListSyntax GetApplicableList(SyntaxNode node);
+        protected abstract SeparatedSyntaxList<TListItemSyntax> GetListItems(TListSyntax listSyntax);
+        protected abstract bool PositionIsApplicable(
+            SyntaxNode root, int position, SyntaxNode declaration, TListSyntax listSyntax);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var span = context.Span;
+            if (!span.IsEmpty)
+            {
+                return;
+            }
+
+            var position = span.Start;
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var generator = document.GetLanguageService<SyntaxGenerator>();
+
+            var declaration = token.Parent.Ancestors().FirstOrDefault(n => GetApplicableList(n) != null);
+            if (declaration == null)
+            {
+                return;
+            }
+
+            var listSyntax = GetApplicableList(declaration);
+            // Make sure we don't have any syntax errors here.  Don't want to format if we don't
+            // really understand what's going on.
+            if (listSyntax.GetDiagnostics().Any(d => d.Severity == DiagnosticSeverity.Error))
+            {
+                return;
+            }
+
+            if (!PositionIsApplicable(root, position, declaration, listSyntax))
+            {
+                return;
+            }
+
+            var listItems = GetListItems(listSyntax);
+            if (listItems.Count <= 1)
```

Comment:
I would find this feature useful even for 1 parameter if it's very long combined with the method name. Often I have to wrap right after typing the first parameter. Why not offer for 1 parameter too given that we're only offering this if the line is long anyway?

Reply:
I found it very noisy, and decided it was nicer to just avoid that scenario and say it's good enough to have people just do that manually.

Reply:
@sharwell for your thoughts here as well.  One of those cases where something is valid, but may be too noisy.

Reply:
I think we should offer for one item when multi-line items are supported:

```
Method(() =>
{
});

Method(
    () =>
    {
    });
```

Reply:
i'd be interested in persuing this.  But i actually think i would need some design help here.  There is a bunch of unknowns and complexity around multi-line constructs, and it's not clear to me (having not spent too cycles on it) what a suitable wrapping strategy is that doesn't mess up code.

In your experience with style-cop, did you work in this problem space?  If so, can you share any strategy tips on how to deal with this situation  If so, i'd be happy to apply it in the future, which might also mean we could relax other restrictions here i've placed here because i felt this was too complex to handle. 

Thanks!